### ISSUE TITLE
Detection and mitigation of deleted extension versions could be improved

### ISSUE THREAD

**Issue by lagalbra** [Azure/WALinuxAgent#959]
There are references to extension versions `1.4.*.*` in `/var/lib/waagent/Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux.104.manifest.xml` and older, but not in `/var/lib/waagent/Microsoft.EnterpriseCloud.Monitoring.OmsAgentForLinux.105.manifest.xml` and newer. It seems the extension publisher has deleted these versions, and basically we do not support that scenario in the agent (we always ask publishers to never remove the latest version).

In this particular case, the issue manifested when a VM repeatedly gave "Failed to load manifest file" errors in `/var/log/waagent.log` approximately every 8 seconds.

The agent should be able to deal with this if the customer stops the agent, removes any reference of this version from disk (if the files are complete, the agent is already doing this) and then restarts the agent. However, perhaps this could be more hardened and automatic in the agent.

More details can be found in the incident 52697262; Hans and lagalbra have context.

**Comment by NarineM** [comment]
Hi Laura,
OMS Linux team has not removed previous versions of OMs linux extension. There are customer VMs that are running old versions. I can provide replication status logs if needed. Is it possible that this is caused by something else?

**Comment by lagalbra** [comment]
It does appear as though versions 1.4.x.x of the OmsAgentForLinux extension are still published and have not been removed:

```
> Get-AzureRmVMExtensionImage -Location WestUS -PublisherName Microsoft.EnterpriseCloud.Monitoring -Type OmsAgentForLinux | Select Type, Version

Type             Version
----             -------
OmsAgentForLinux 1.0.217.0
OmsAgentForLinux 1.2.148.0
OmsAgentForLinux 1.3.127.5
OmsAgentForLinux 1.3.127.7
OmsAgentForLinux 1.3.18.7
OmsAgentForLinux 1.4.45.2
OmsAgentForLinux 1.4.45.3
OmsAgentForLinux 1.4.55.4
OmsAgentForLinux 1.4.56.5
```

**Comment by boumenot** [comment]
@hglkrijger / @lagalbra - can you provide more context on this issue.  

 1. Do you know if this has been addressed with 1188daad0e0cc32aecd0f0048d3c4cb0a1db1276?  All of these files are moved when the incarnation number rolls over.  
 1. I can't tell if this is a duplicate of #840.

I also wonder if this is related to the caching bug we introduced with 2.2.2[0123].

**Comment by hglkrijger** [comment]
No, this is an issue where:

- version _n_ of an extension is installed via incarnation _i_
- we receive new manifests via incarnation _i+1_
- version _n_ has been removed from the manifest, and _n+1_ has been added
- the agent tries to obtain the disable/uninstall commands from the _i+1_ manifest for _n_ but they do not exist any more; this repeats

If the agent is restarted this problem goes away because we have to rebuild the references to the extensions on disk (as I recall). So the idea here was to handle this scenario more gracefully if extensions are unpublished (or are not in the manifest for whatever reason).
